### PR TITLE
Use proto file name to infer message type

### PR DESCRIPTION
**Pull request by honnix**
# TL;DR
Use proto file name to infer message type to have more precise unmarshalling.

## Type
- [x] Bug Fix
- [ ] Feature
- [ ] Plugin

## Are all requirements met?

- [x] Code completed
- [x] Smoke tested
- [x] Unit tests added (the existing unit tests cover this change well)
- [ ] Code documentation added
- [ ] Any pending items have an associated Issue

## Complete description
flytekit generated proto file name has a pattern to precisely indicate the message type: `.*_1.pb`  is task, `.*_2.pb` is workflow and `.*_3.pb` is launch plan. We can rely on this pattern to do more precise unmarshalling. Note that we still try other types if the unmarshalling based on the precisely matched type does not work.

This PR keeps backward compatibility so if the pattern is not seen,  it goes through all possible types and tries one by one.

## Tracking Issue
fixes https://github.com/flyteorg/flyte/issues/3042

## Follow-up issue
_NA_

### CODE REVIEW

File: cmd/register/register_util.go [hunk 1]
```diff
@@ -92,37 +92,72 @@ var projectColumns = []printer.Column{
...
+func reorderUnMarshalFuncs(fname string) []unMarshalFunc {
+	if match := fnameRegex.FindStringSubmatch(fname); match != nil {
+		indexStr := match[fnameRegex.SubexpIndex("index")]
+		index, err := strconv.Atoi(indexStr)
+		if err != nil {
+			panic(fmt.Sprintf("unexpected error when coverting [%s] to int, file name [%s]", indexStr, fname))
+		}
+
+		var reordered []unMarshalFunc
+		for i, f := range unMarshalFuncs {
```
Comment (honnix) [inline cmd/register/register_util.go:150]: This loop basically bubbles up the matched function to the first place.

File: cmd/register/register_util.go [hunk 2]
```diff
@@ -92,37 +92,72 @@ var projectColumns = []printer.Column{
 	{Header: "Additional Info", JSONPath: "$.Info"},
 }
 
+var fnameRegex = regexp.MustCompile(`^.*_(?P<index>[1-3])\.pb$`)
```
Comment (pingsutw) [inline cmd/register/register_util.go:95]: nit: we could add some comments here

Comment (honnix) [reply cmd/register/register_util.go:97]: Added in a latter commit.

**Comment by honnix** [comment]
Thank you for approving. Could you please help merge it when ready? Thanks.
